1-Draft/RFCNNNN-Side-By-Side-Support.md

```diff
@@ -0,0 +1,253 @@
+
+The end goal is for users to be easily able to do any of the following with a
+PowerShell installer:
+
+* install the latest version of PowerShell side-by-side with an existing
+version from a UI (e.g. `.msi`, `.pkg`, `.deb`)
+* install the latest version of PowerShell side-by-side with an existing
+version from the command line (e.g. `sudo apt-get install ...`, `brew cask
+install ...`, `choco install ...`)
+
+_Note: I'm a little uncertain of these details here, so please correct me where
+appropriate._
+
+The current release process seems to include updated published versions of
+packages in the various package repositories (for `apt-get`, homebrew, and
+Chocolatey). The problem with this approach is that it does not support
+side-by-side installations at the command line. For example, if I have
+PowerShell Core 6.2.0 installed on macOS and I want to install PowerShell Core
+6.2.1 side-by-side, I cannot. I can upgrade my installed version, reinstall
+the current version, or install a preview version (since that comes from a
+different package name), but I cannot install a different version side-by-side.
+
+Instead, if the release process not only updated the latest package but also
+created a version-specific package in the various repositories, then users
+could specify a specific version that they want to install (e.g. `brew cask
+install powershell@6.2.0`) to install different versions of PowerShell
+side-by-side with one another, with the versions automatically being installed
+in an appropriate version folder (which they are already anyway). Chocolatey
+already has support for installing a specific version, but it does not seem to
+work at the moment (I tried installing PowerShell Core 6.2.0 and ended up with
+6.2.1, with an error indicating that 6.2.0 was not installed).
+
+The UI installers should provide options to upgrade the latest installed
+version of PowerShell or install a new version side-by-side.
+
+### Listing installed versions of PowerShell
+
+```bash
+# This launches the latest version of PowerShell, returns a list of installed
+# versions and their installation location, and then exits
+pwsh -ListVersions
+# This does the same thing
+pwsh -lv
+```
+
+```output
+6.2.1 [/usr/local/microsoft/powershell/6.2.1/]
+7.0.100-preview1 [/usr/local/microsoft/powershell/7.0.100-preview1]
+```
+
+```powershell
+
+# This returns a list of installed versions and their installation location as
+# objects
+Get-InstalledPSVersion
+```
+
+```output
+Version                    InstallLocation
+-------                    ---------------
+6.2.1                      C:\ProgramFiles\PowerShell\6.2.1
+7.0.100-preview1           C:\ProgramFiles\PowerShell\7.0.100-preview1
+```
+
+### Launching a specific version of PowerShell
+
+```bash
+# This launches the latest version of PowerShell, gets the installation
+# location of the specified version, launches pwsh for that version, and exits
+# when the launched version exits
+pwsh -Version 6.2.1
```

Comment:
There are some implications we need to think about here.  On Unix systems, we can leverage `exec` so that there is a single pwsh process.  On Windows, if you run this from cmd or PowerShell, the assumption is you want to use the current console window.  So that means that `pwsh` will start the newest one (as that's likely the one in PATH), which will then start a child process with the requested version.  So you'll end up with two pwsh processes.  There might be a way for `pwsh.exe` to host the requested `pwsh.dll`, but then you might have issues with `pwsh.exe` being compiled against a different version of .NET Core.  Need to think through this or just accept there's two processes.

Reply:
Within PowerShell at least, couldn't it intercept this request to start the native app called `pwsh` with a specific `-Version` param/arg, lookup that version and start it, passing it the rest of parameters except for `-Version`?  I'm not a fan of special-case code like this but it seems doable.

Reply:
It could be special cased in PowerShell itself, but not sure if it's worth it.  Also, it seems that this is most useful for scripts using `#Requires -Version X` to automatically run under those versions which isn't called out in this RFC.

Reply:
The “PowerShell.exe -Version x.x -File “C:\Path\To\script.ps1” could work here as well. It would be nice if a warning could be displayed, if that version isn’t installed and include what version was used instead.

Reply:
Also, I think it would be good to spawn a new child process, if calling a different  “pwsh” from the console. This is expected behavior in Windows and *nix world when requesting a different shell. Unless a very good cleanup job was done of the current process, using some .dll swapping could get messy.

Reply:
I think there should be an error if the specified version isn't available, not using something else.

Reply:
What if the version is not installed? Will we fail or run upper version?

Reply:
It seems that if a specific version is requested and not found, it should not run and would return an error.  If `6.2` is requested, then 6.2.2 would run, if available.  It probably doesn't make sense to support any `preview` releases with this feature.